Runtime Version Scanner: Display only a publisher's content

toph-allen opened the pull request:
This PR adds the following features:

- Displays only visitor-scoped content. Currently, it still uses the `v1/content` API, so behind the scenes admin users still see all content downloaded and have it scoped client-side to their owned/collaborating content. However, this is factored in the app such that it'd be easy to replace it with a call to the `v1/search/content` endpoint.
- Uses a visitor-scoped Connect client. This is added as a Shiny module, moving all but two lines of code out of the main app (in the interests of clarity and exploring how we might better provide this functionality to our users).
- Only displays content that uses selected EOL runtimes. Table highlights the matching runtime, in case the content uses more than one runtime and only one is EOL.
- Displays summary text at the top about the selected EOL runtimes and the number of matching pieces of content.

The app is deployed on Dogfood and connect.posit.it:
https://dogfood.team.pct.posit.it/runtime-version-scanner/
https://connect.posit.it/runtime-version-scanner/

@jonkeane I mentioned that I felt like I had a bit more clarity about when the app was ready for feedback — I think the changes made here make the app fit the top-level brief [here](https://github.com/posit-dev/connect/issues/29629) (although that does need refinement, I'll get to that later). I feel like the functionality here could spark opinions about what people would _actually_ want to see.

---- 

Updates after feedback

- Displays the entire content data frame when nothing is selected for filtering.
- Uses sliders to display filter controls.
- Renamed from "EOL Runtime Scanner" to "Runtime Version Scanner"
- Adds tests

toph-allen commented on extensions/eol-runtime-scanner/app.R:
```diff
@@ -6,14 +6,38 @@ library(dplyr)
   spinner.color = "#7494b1"
 )
 
+app_mode_groups <- list(
+  "API" = c("api", "python-fastapi", "python-api", "tensorflow-saved-model"),
+  "Application" = c(
+    "shiny",
+    "python-shiny",
+    "python-dash",
+    "python-gradio",
+    "python-streamlit",
+    "python-bokeh"
+  ),
+  "Jupyter" = c("jupyter-static", "jupyter-voila"),
+  "Quarto" = c("quarto-shiny", "quarto-static"),
+  "R Markdown" = c("rmd-shiny", "rmd-static"),
+  "Pin" = c("pin"),
+  "Other" = c("unknown")
+)
+
+app_mode_lookup <- with(
+  stack(app_mode_groups),
+  setNames(as.character(ind), values)
+)
```
Totally agree, I copied and pasted this from the Usage Metrics Dashboard, and that it seems to be becoming the go-to way to show app modes to users does suggest that it should be… somewhere!

When you say "somewhere in the API we could pull it from," my first thought was an endpoint that returned… data like _this_, i.e. a dictionary of app mode groups, but I also wondered if you meant, like, an additional field on content objects. But having the full list would be good.

I don't think we have endpoints in the Connect API yet which provide… like, ontological/schema-like info like this, so this would be a first.

It would also be reasonable to just put this object or something like it in `connectapi`! :)

I'll file an issue on `connectapi` at the very least! https://github.com/posit-dev/connectapi/issues/419

jonkeane replied:
> When you say "somewhere in the API we could pull it from," my first thought was an endpoint that returned… data like this, i.e. a dictionary of app mode groups, but I also wondered if you meant, like, an additional field on content objects. But having the full list would be good.

Yeah, in general it's nice for APIs to be self describing, especially when there are peculiarities like there are for this one. But if we don't have that (yet) or decide as maintainers of both the server code and the client libraries that we would rather have them in the client libraries, that's fine too!

toph-allen commented on extensions/eol-runtime-scanner/app.R:
```diff
@@ -107,38 +171,61 @@ server <- function(input, output, session) {
     reactable(
       data,
-      defaultPageSize = 25,
+      pagination = FALSE,
```
Because `reactable` is generally pretty fast to render large tables, user-scoped content tables aren't _that_ large, and sometimes paginating a table is a worse experience than just scrolling a medium-long table.

jonkeane replied:
I would leave it as paginated, maybe with a large number of items per page. But "user-scoped content tables aren't that large" isn't actually something we can (ever) know if this is totally true for. On some servers, for some people, it might have 10 things, on other servers, for other people it might have 1000 things. Being resilient to that would be good (or at least, I don't think we should _remove the resilience_ unless there's a really good reason)

toph-allen commented on extensions/eol-runtime-scanner/app.R:
```diff
@@ -107,38 +171,61 @@ server <- function(input, output, session) {
+    pv <- input$py_versions
+    qv <- input$quarto_versions
+
+    if (length(rv) == 0 & length(pv) == 0 & length(qv) == 0) {
+      paste0(
+        "Please select the versions of R, Python, and Quarto you ",
+        "wish to scan for."
+      )
```
I felt this way at first, but using that version felt worse to me for reasons described in my comment. But I'm interested to see the feedback we get.

Like, it feels weird to suddenly hide all of the other content when you get into a "content selected" mode. I almost feel like I (Toph) would want to have a toggle between a "show only matching content" and "show all content but highlight matching content" modes. Showing the full table in the zero-selected state and then hiding non-matching content when you make a selection feels kinda weird.

Oh! I think it's because the language filters operate in an "or" mode — so as you add additional filter, the scope of things that match expand, which is different from how a lot of table filters traditionally work but makes sense for this use case. (Like, if I select R 3.6.4 and Python 3.7.0, I clearly want to see content that uses _either_ of them, not _just_ content that uses _both_.)

I don't want to over-index on my own preferences, but since this is perhaps the less-obvious (and thus perhaps wrong!!) choice for the interaction, I'm… kinda interested to show it to SE.

jonkeane replied:
> I don't want to over-index on my own preferences, but since this is perhaps the less-obvious (and thus perhaps wrong!!) choice for the interaction, I'm… kinda interested to show it to SE.

Yup, we should not debate this here and get ossified in our thinking. We should get feedback on it. But the first view showing nothing _looks broken_. Is the point like you're saying. Giving someone ideas bout what's there and not "oh hmm this must be busted the list is empty" helps keep people engaged.

dotNomad commented:
I will give my two cents here about the current state of this:

I didn't mind the "No rows found" start. I do think it would be improved if it started with all of my content though since that may give me a baseline to start with the filters. For example, if I have a lot of `4.3.1` R Version content maybe I set that as my minimum version I want to filter for to check for old language versions.

I still find it strange that we are calling with an End-of-Life Runtime Scanner, but we aren't telling users what is EOL.

> We call this the EOL enumerator in the issue, but I think we want something slightly more generic as the name in the app. This is more general, it might be looking for things that depend on "old" versions even if they are not end of life yet. We can (and should) put EOL in the description as a "one thing you might do with this is..." but we don't need to have it plastered in the title

@jonkeane described this a bit here, and I do think that is accurate to how I'm thinking about this too.

My expectation is that I set a minimum myself, but the tool would help me figure out that minimum. For example the dashboard should give me a clear idea that Python `3.8` is EOL, but not limit me to that selection. Python `3.12` ended active support so maybe I want to be proactive and set my minimum Python version to `3.13`. Ideally the dashboard should support both flows. The key differences there are:
- a minimum version rather than selecting versions (a slider is a good idea or a select field specifying a minimum)
- having something in that selection telling me what is actually EOL

toph-allen commented on extensions/eol-runtime-scanner/app.R:
```diff
@@ -6,82 +6,190 @@ library(dplyr)
+
+# Functions to handle versions
+
+# Takes a column from a data frame and returns a character vector sorted as
+# version numbers.
+ordered_version_levels <- function(versions) {
+  rv <- versions |>
+    na.omit() |>
+    unique() |>
+    as.numeric_version() |>
+    sort() |>
+    as.character()
+}
```
Yep!

toph-allen commented:
I added tests for the version ordering functions. They're pretty basic, but I think they cover most cases (we don't need to be too defensive since we expect _relatively_ clean data from Connect, and they aren't used anywhere but here).

I don't actually have `library(testthat)` in the test script, and this ensures that `renv` does not add it to our app's dependencies.

I also added a `Makefile` with a target — @cgraham-rs perhaps a hook like this can be used when adding a version in the future.

jonkeane commented on extensions/eol-runtime-scanner/tests/test_version_ordering.R:
```diff
@@ -0,0 +1,69 @@
+  )
+
+  # The input vector must already be sanitized.
+  expect_error(sort_unique_versions(versions))
```
Could you assert some portion of the error you get here (so that we know we are getting the one we expect)?

jonkeane reviewed (commented):
A few comments on the tests. I'm ok to merge this after you address the test comments and get feedback (or get feedback without merging, though small targeted changes are _so much nicer_ to review I would bias towards getting to that if I were you).

toph-allen commented:
> A few comments on the tests. I'm ok to merge this after you address the test comments and get feedback (or get feedback without merging, though small targeted changes are _so much nicer_ to review I would bias towards getting to that if I were you).

Yeah, I'm eager to get this merged so that my next changes are in a new diff; I'll merge after addressing your test comments. Thank you!

toph-allen commented:
> I'm ok to merge this after you address the test comments and get feedback (or get feedback without merging, though small targeted changes are so much nicer to review I would bias towards getting to that if I were you)

Going to merge without an approval per the verbal approval in @jonkeane’s comment here. A version of this app which is functionally identical to this was already posted to the appropriate channel for feedback, and I'm going to collate that feedback and respond to it & add new functionality in a subsequent pull request.
